File: src/main/java/com/cisco/td/general/cocroach/TabularExchangeRateProvider.java
package com.cisco.td.general.cocroach;

import com.google.common.collect.ImmutableSortedMap;
import org.joda.time.LocalDate;

import java.util.Map;
import java.util.NavigableMap;

public class TabularExchangeRateProvider implements ExchangeRateProvider {
    private final NavigableMap<LocalDate, Double> knownRates;

    public TabularExchangeRateProvider(Map<LocalDate, Double> knownRates) {
        this.knownRates = ImmutableSortedMap.copyOf(knownRates);
    }

    public static TabularExchangeRateProvider hardcoded() {
        return new ExchangeRatesReader().parse(
                ByteSources.fromResource(TabularExchangeRateProvider.class, "rates_2021.txt"),
                ByteSources.fromResource(TabularExchangeRateProvider.class, "rates_2022_a.txt"),
                ByteSources.fromResource(TabularExchangeRateProvider.class, "rates_2022_b.txt"),
                ByteSources.fromResource(TabularExchangeRateProvider.class, "rates_2023.txt")
        );
    }

    @Override
    public double rateAt(LocalDate day) {
        Map.Entry<LocalDate, Double> maybeEntry = knownRates.floorEntry(day);
        if (maybeEntry != null) {
            return maybeEntry.getValue();
        } else {
            throw new IllegalArgumentException("can not find rate for " + day);
        }
    }
}

File: src/main/kotlin/cz/solutions/cockroach/TabularExchangeRateProvider.kt
package cz.solutions.cockroach

import org.joda.time.LocalDate
import java.util.*

class TabularExchangeRateProvider(
    knownRates: Map<LocalDate, Map<Currency, Double>>
) : ExchangeRateProvider {
    private val knownRates: NavigableMap<LocalDate, Map<Currency, Double>> = TreeMap(knownRates)

    companion object {
        fun hardcoded(): TabularExchangeRateProvider {
            return fromSource(ClasspathCnbYearRatesSource(), 2021..2025)
        }

        fun fromSource(source: CnbYearRatesSource, years: IntRange): TabularExchangeRateProvider {
            val chunks = years.flatMap { source.loadYear(it) }
            return ExchangeRatesReader.parse(*chunks.toTypedArray())
        }
    }

    override fun rateAt(day: LocalDate, currency: Currency): Double {
        if (currency == Currency.CZK) return 1.0
        val perCurrency = knownRates.floorEntry(day)?.value
            ?: throw IllegalArgumentException("can not find rate for $day")
        return perCurrency[currency]
            ?: throw IllegalArgumentException("can not find rate for $day in $currency")
    }
}